[NestedTensor] Integrate sum along the jagged dimension into NestedTensor

jananisriram opened the pull request:
Summary: Modify the existing `sum` operator in PyTorch, invoked by `torch.sum`, to allow for reductions along the ragged dimension of a nested tensor. This diff enables PyTorch users to invoke `torch.sum` on a nested tensor with `dim=1`, where `ragged_idx=1`.

Functions modified in `caffe2/torch/nested/_internal/ops.py`:
- `sum_dim_IntList()`: The function assumes that `ragged_idx=1`; in the case that `dim=1` as well, where `dim` is the dimension on which we reduce, this diff invokes the PyTorch benchmark found in D58423489. Specifically, this diff pads a nested tensor, e.g. of logical shape `(B, *, M)`, using [`torch.ops.aten._jagged_to_padded_dense_forward`](https://www.internalfb.com/code/fbsource/[92c2a067ab04e3eebc999254fed4ae2fbea6def3]/fbcode/deeplearning/fbgemm/fbgemm_gpu/fb/inductor_lowerings/elementwise_ops.py?lines=26), then reduces across the `*` dimension (`dim == 1`) to a `(B, M)` output tensor.
- `_wrap_jagged_dims()`: This diff adds special handling to allow for the case where `dim` contains `1` and not `0`, but to continue disallowing the case where `dim` contains `0` and not `1`. In this function's creation, I created a helper function, `_get_condition_for_invalid_jagged_reductions()`, which makes it clearer which conditions apply to which operators. Specifically, operators which are enabled with jagged reductions are specified at the top of the file in `SUPPORTED_JAGGED_REDUCTIONS` and have a different set of conditions that need to be tested, as reducing along `dim == 1` without `dim == 0` is now possible.

Functions modified in `caffe2/test/test_nestedtensor.py`:
- `test_sum_int_DimList()`: This diff adds special handling in the `sum` unit test to allow for the case where `dim` contains `1` and not `0`, but to continue disallowing the case where `dim` contains `0` and not `1`.
- `test_sum_int_DimList_ragged_dim_1()`: This diff adds a new unit test which verifies the accuracy and feasibility of reducing along the jagged dimension of a nested tensor.

Notes:
- This diff solely adds functionality for the case in which we reduce only along the ragged dimension. Cases in which we reduce along both the ragged and another dimension, like `dim == (1, 2)`, are not permitted, as this set of diffs focuses primarily on the former.
- The `sum` operator is the only operator which uses the function `_wrap_jagged_dims()`; all other operators use `_wrap_jagged_dim()`. I would like to later look into why this is the case and if we can consolidate this!
- I modified some of the comments in the `sum` function as well as the unit tests for more clarity.

Test Plan:
Verify that existing (`test_sum_int_DimList`) and new (`test_sum_int_DimList_ragged_dim_1`) unit tests pass via the following command:

```
buck2 run mode/{opt,inplace} //caffe2/test:nested -- --regex test_sum_int_DimList
```

Differential Revision: D59571209

davidberard98 reviewed (changes requested):
I see some lint & test failures - can you take a look at those?

for test failures, if you haven't added full support for some scenarios, throw exception for those scenarios - and either don't test those scenarios, or assert that exceptions are thrown

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -46,16 +47,30 @@ def _wrap_jagged_dims(ndim, dims, op_name):
-        raise RuntimeError(
-            f"{op_name}(): applying over the {apply} dimension, but not the {not_apply}"
-            " dimension is not supported for NestedTensor"
+    if op_name in SUPPORTED_JAGGED_REDUCTIONS:
```
nit: instead of maintaining a separate list of `SUPPORTED_JAGGED_REDUCTIONS`, I think I'd prefer one or two kwargs conceptually customizing `_wrap_jagged_dim()`'s behavior. I think the cases we want to support are:
1. pre-PR case: allow everything except `batch but not ragged` or `ragged but not batch` dims
2. new case: allow everything except batch-only

I think we can come up with a nice set of kwargs that will cover those cases :)

Another suggestion is to perhaps have `_wrap_jagged_dim()` return `(wrapped dim, operating_on_batch, operating_on_ragged)` to provide flexibility to the caller on how to handle these. as-is, the return values aren't super intuitive

jananisriram replied:
agreed, and i think this follows a point made in a future comment as well! i've made some modifications to the return values from `_wrap_jagged_dims` to make the logic clearer and have also cleaned up the implementation of `sum` for `NestedTensor`

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -46,16 +47,30 @@ def _wrap_jagged_dims(ndim, dims, op_name):
     # map to inner dims so we can print a nicer error message.
     zero_in_dims = 0 in wrapped_dims
     one_in_dims = 1 in wrapped_dims
```
I realize this is a preexisting condition, but imo it'd be better to have conceptual names here (e.g. `operates_on_ragged`, `operates_on_batch`). It looks like the names here hide the fact that the logic doesn't take into account when the NJT's ragged_idx is not 1.

not necessarily your problem to fix, but if you wanted to refactor this slightly as you expand it, I wouldn't frown upon it :)

jananisriram replied:
completely agree with this! i was planning on adding a new diff to refactor this (so as to not both add new logic and change existing variable names in one pr), but i also think it makes more sense to refactor as i go. thanks!

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -842,15 +857,35 @@ def sum_dim_IntList(func, *args, **kwargs):
+
+    if (
+        0 not in new_kwargs["dim"]
+        and 1 in new_kwargs["dim"]
```
as mentioned above, I think we need to take into account NJTs with ragged_idx != 1

jananisriram replied:
agreed, and this is actually going to be added in a later diff! for now, as discussed above, i'll refactor these names to be more explicit/clear, and modify the logic here when i add support for NJTs with ragged_idx != 1

jbschlosser reviewed (commented):
Thanks for the PR! It'll be awesome to have this in here :) added some minor nits on structure / testing but overall it's looking good! Most of my comments here should apply to the `mean` PR as well.

Definitely agree with David about being explicit on what is supported and not, explicitly throwing nice errors for the unsupported set. Ideally, we also want comprehensive tests for all these, including the error cases (assert that the errors are thrown here).

jananisriram commented on test/test_nestedtensor.py:
```diff
@@ -4051,6 +4061,39 @@ def test_sum_int_DimList(self, device, keepdim):
+                [torch.sum(t, dim=0).unsqueeze(0) for t in nt.unbind()]
+            )
+
+            self.assertNotIsInstance(
+                out_actual,
+                NestedTensor,
+                "sum(): the result of reducing a nested tensor along the ragged dimension is a dense tensor",
+            )  # output is a dense tensor
```
thanks! this seems a bit clearer :)

jananisriram commented on torch/nested/_internal/ops.py:
```diff
@@ -842,15 +857,35 @@ def sum_dim_IntList(func, *args, **kwargs):
     )
     inp = new_kwargs.pop("input")
     assert inp._ragged_idx == 1
```
good point, thanks! also added a new test to check that this error is thrown

jananisriram commented on test/test_nestedtensor.py:
```diff
@@ -4051,6 +4061,39 @@ def test_sum_int_DimList(self, device, keepdim):
+        tensor_lists = self._get_example_tensor_lists(
+            include_list_of_lists=False, include_requires_grad=components_require_grad
+        )
+        reduce_dim = (1,)
```
added a set of tests that cover the possible reduction dimensions (combinations of batch, ragged, non-batch), as well as `ragged_idx != 1`. thanks for the suggestion!

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -847,28 +826,75 @@ def is_same_size_default(func, *args, **kwargs):
     )
 
-    if not ragged_reduced_away:
+    # invalid reductions
+    batch = reduce_on_batch and not reduce_on_ragged and not reduce_on_non_batch
+    batch_and_non_batch = (
+        reduce_on_batch and not reduce_on_ragged and reduce_on_non_batch
+    )
+    ragged_and_non_batch = (
+        not reduce_on_batch and reduce_on_ragged and reduce_on_non_batch
+    )
+
+    if batch or batch_and_non_batch:
+        raise RuntimeError(
+            "sum(): not supported along a batch dimension for NestedTensor"
+        )
+
+    if ragged_and_non_batch:
+        raise RuntimeError(
+            "sum(): not supported along a ragged and non-batch dimension for NestedTensor"
+        )
```
nit: I find this a bit tricky to read, but it seems like these are the cases we care about:
1. raggedness is reduced away -> return dense
    * not reducing over batch dim -> need to read offsets to determine batch items
        * error for non-batch as unsupported
        * call `_jagged_to_padded_dense_forward()` otherwise
    * reducing over batch dim -> no need to read offsets
        * call `func` directly on `values` (NB: avoid duplicate dims in the dim_list arg)
2. raggedness is not reduced away -> return NJT
    * error if trying to reduce over the batch dim
    * call `func` directly on `values`

maybe structuring like this (with comments) could make it easier to follow, wdyt?

jananisriram replied:
i agree, i think this is a good point on readability! refactored my code to make it easier to understand, with comments clearly explaining what is currently supported and what isn't. thank you!

jbschlosser reviewed (commented):
Thanks for your work on refactoring `_wrap_jagged_dims()`! It's tedious I know but I appreciate the effort getting this to a good state :) some more minor comments below

jananisriram commented on torch/nested/_internal/ops.py:
```diff
@@ -36,26 +36,20 @@ def _wrap_jagged_dim(
+
+    reduce_on_batch = 0 in wrapped_dims
+    reduce_on_ragged = ragged_idx in wrapped_dims
+    reduce_on_non_batch = any(n in wrapped_dims for n in range(2, ndim))
```
thanks for catching this so that i can make the code easier to modify for a later PR (for `ragged_idx != 1`)!

jananisriram commented on torch/nested/_internal/ops.py:
```diff
@@ -36,26 +36,20 @@ def _wrap_jagged_dim(
+    reduce_on_ragged = ragged_idx in wrapped_dims
+    reduce_on_non_batch = any(n in wrapped_dims for n in range(2, ndim))
+
+    outer_to_inner_dim = tuple(_outer_to_inner_dim(ndim, d) for d in dims if d != 0)
```
oops thanks for the catch!

jananisriram commented on test/test_nestedtensor.py:
```diff
@@ -3761,11 +3762,7 @@ def test_split_with_sizes(self, device):
 
     def test_softmax(self, device):
         nt = random_nt_from_dims(
-            [3, None, 5],
```
had some github merge issues, thanks for pointing this out! re-exporting and fixing this now

jananisriram commented on torch/nested/_internal/ops.py:
```diff
@@ -515,21 +509,6 @@ def _softmax_default(func, *args, **kwargs):
     return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
 
 
-@register_jagged_func(
-    torch.ops.aten._softmax_backward_data.default,
-    "grad_output: jt, output: jt, dim: any, input_dtype: any",
-)
-def _softmax_backward(func, *args, **kwargs):
-    _, new_kwargs = normalize_function(
-        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
-    )
-    grad_out = new_kwargs.pop("grad_output")
-    output = new_kwargs.pop("output")
-    return NestedTensor(
-        func(grad_out._values, output._values, **new_kwargs), **extract_kwargs(grad_out)
-    )
```
see [here](https://github.com/pytorch/pytorch/pull/130425/files/96f1601591c895c30f0f8b47771772464a1bbdd0#r1679892793)

jananisriram commented on test/test_nestedtensor.py:
```diff
@@ -5220,15 +5415,6 @@ def test_sdpa(self, device, dtype):
         d1_grads = torch.autograd.grad(attn_d1.sum(), (q_d1, k_d1, v_d1))
         d2_grads = torch.autograd.grad(attn_d2.sum(), (q_d2, k_d2, v_d2))
 
-        # Simple case 3: batch_size = 1, seq_len = 1
-        q_3 = torch.randn(1, 8, 16, dtype=dtype, device=device)
-        q_nt_3 = torch.nested.as_nested_tensor([q_3], layout=torch.jagged)
-        q_nt_3 = q_nt_3.transpose(1, 2)
-        attn_out = torch.nn.functional.scaled_dot_product_attention(
-            q_nt_3, q_nt_3, q_nt_3
-        )
-        self.assertEqual(attn_out.shape, q_nt_3.shape)
```
see [here](https://github.com/pytorch/pytorch/pull/130425/files/96f1601591c895c30f0f8b47771772464a1bbdd0#r1679892793)

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+
+    if reduce_on_ragged and inp._lengths is not None:
+        raise RuntimeError(
+            "sum(): not supported for nested tensors with holes, i.e. when lengths is not None, "
```
nit: maybe just say "not supported for nested tensors where lengths is not None if reducing across the ragged dimension for Nested Tensor"

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+            reduce_on_batch
+        ):  # invalid reduction cases: (batch), (batch, non-batch), etc.
+            raise RuntimeError(
+                "sum(): not supported along a batch dimension for NestedTensor"
```
nit: maybe also mention that reduction is supported when batch dim + ragged dim are reduced

jananisriram replied:
i do mention this in a [comment](https://github.com/pytorch/pytorch/pull/130425/files#diff-5ab78b751bd84fe009dbdb8899eb434a0278315bc33f15dfe45d8d87a2afa662R883) in my code; however, would it be clearer to modify this to `sum(): not supported along only the batch dimension for NestedTensor`?

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+                dim=inp._ragged_idx,
+            )  # need to read offsets --> pad jagged dimension and apply sum
+
         if new_kwargs["keepdim"]:
             out = out.unsqueeze(0)
```
note: I don't think this is actually right? e.g. you'll end up with a size of (1, ...) which I don't think matches the semantics of normal keepdim; I think you'd usually expect the output shape to be the same as the input shape (realize this is pre-existing before your PR)

maybe can you open a github issue for this with some details to track this?

davidberard98 replied:
offline discussion: this is actually fine, doesn't need an issue

jananisriram replied:
resolved offline! the pytorch documentation says that intended functionality for `keepdim=True` is for the number of dimensions in the input and output to match, rather than the shapes themselves to match

```
If keepdim is True, the output tensor is of the same size as input except in the dimension(s) dim where it is of size 1.
```

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+            + "if reducing across the ragged dimension for NestedTensor"
+        )
+
+    if reduce_on_ragged:  # raggedness reduced away --> return dense tensor
```
minor nit: the branching here is a bit confusing IMO, but I also don't really have a suggestion on how to make it cleaner haha

jananisriram replied:
agreed! i think it was worse before and now it's a bit more cleaned up, but it's still tedious to read unfortunately. one possible solution is to create individual boolean conditions for each reduction case, but this could also make the code long-winded to read

davidberard98 reviewed (approved):
a few nits, otherwise LGTM!

jananisriram commented on torch/nested/_internal/ops.py:
```diff
@@ -36,27 +36,23 @@ def _wrap_jagged_dim(
+
+    operate_on_batch = 0 in wrapped_dims
+    operate_on_ragged = ragged_idx in wrapped_dims
+    operate_on_non_batch = any(d for d in wrapped_dims if d != 0 and d != ragged_idx)
```
thanks for the catch! i think this was the intent but it got lost along the way. will fix this!

jananisriram committed:
[NestedTensor] Integrate sum along the jagged dimension into NestedTensor (#130425)

Summary:
Modify the existing `sum` operator in PyTorch, invoked by `torch.sum`, to allow for reductions along the ragged dimension of a nested tensor. This diff enables PyTorch users to invoke `torch.sum` on a nested tensor with `dim=1`, where `ragged_idx=1`.

Functions modified in `caffe2/torch/nested/_internal/ops.py`:
- `sum_dim_IntList()`: The function assumes that `ragged_idx=1`; in the case that `dim=1` as well, where `dim` is the dimension on which we reduce, this diff invokes the PyTorch benchmark found in D58423489. Specifically, this diff pads a nested tensor, e.g. of logical shape `(B, *, M)`, using [`torch.ops.aten._jagged_to_padded_dense_forward`](https://www.internalfb.com/code/fbsource/[92c2a067ab04e3eebc999254fed4ae2fbea6def3]/fbcode/deeplearning/fbgemm/fbgemm_gpu/fb/inductor_lowerings/elementwise_ops.py?lines=26), then reduces across the `*` dimension (`dim == 1`) to a `(B, M)` output tensor.
- `_wrap_jagged_dims()`: This diff adds special handling to allow for the case where `dim` contains `1` and not `0`, but to continue disallowing the case where `dim` contains `0` and not `1`. In this function's creation, I created a helper function, `_get_condition_for_invalid_jagged_reductions()`, which makes it clearer which conditions apply to which operators. Specifically, operators which are enabled with jagged reductions are specified at the top of the file in `SUPPORTED_JAGGED_REDUCTIONS` and have a different set of conditions that need to be tested, as reducing along `dim == 1` without `dim == 0` is now possible.

Functions modified in `caffe2/test/test_nestedtensor.py`:
- `test_sum_int_DimList()`: This diff adds special handling in the `sum` unit test to allow for the case where `dim` contains `1` and not `0`, but to continue disallowing the case where `dim` contains `0` and not `1`.
- `test_sum_int_DimList_ragged_dim_1()`: This diff adds a new unit test which verifies the accuracy and feasibility of reducing along the jagged dimension of a nested tensor.

Notes:
- This diff solely adds functionality for the case in which we reduce only along the ragged dimension. Cases in which we reduce along both the ragged and another dimension, like `dim == (1, 2)`, are not permitted, as this set of diffs focuses primarily on the former.
- The `sum` operator is the only operator which uses the function `_wrap_jagged_dims()`; all other operators use `_wrap_jagged_dim()`. I would like to later look into why this is the case and if we can consolidate this!
- I modified some of the comments in the `sum` function as well as the unit tests for more clarity.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/130425

Test Plan:
Verify that existing (`test_sum_int_DimList`) and new (`test_sum_int_DimList_ragged_dim_1`) unit tests pass via the following command:

```
buck2 run mode/{opt,inplace} //caffe2/test:nested -- --regex test_sum_int_DimList
```

Reviewed By: davidberard98

Differential Revision: D59571209

Pulled By: jananisriram

jananisriram commented:
@pytorchbot merge
